agent/dockerclient/dockerapi/docker_client.go

```diff
@@ -882,38 +883,53 @@ func getMetadataHealthCheck(dockerContainer *docker.Container) apicontainer.Heal
 
 // Listen to the docker event stream for container changes and pass them up
 func (dg *dockerGoClient) ContainerEvents(ctx context.Context) (<-chan DockerContainerChangeEvent, error) {
-	client, err := dg.dockerClient()
+	client, err := dg.sdkDockerClient()
 	if err != nil {
 		return nil, err
 	}
-	dockerEvents := make(chan *docker.APIEvents, dockerEventBufferSize)
-	events := make(chan *docker.APIEvents)
+
+	event := make(chan *events.Message)
 	buffer := NewInfiniteBuffer()
 
-	err = client.AddEventListener(dockerEvents)
-	if err != nil {
-		seelog.Errorf("DockerGoClient: unable to add a docker event listener: %v", err)
-		return nil, err
-	}
-	go func() {
-		<-ctx.Done()
-		client.RemoveEventListener(dockerEvents)
-	}()
+	derivedCtx, cancel := context.WithCancel(ctx)
+	dockerEvents, eventErr := client.Events(derivedCtx, types.EventsOptions{})
+
+	// Cache the event from docker client. Channel closes when an error is passed to eventErr.
+	go buffer.StartListening(derivedCtx, dockerEvents)
+
+	// Receive errors from channels. If error thrown is not EOF, log and reopen channel.
+		go func(){
+			for {
+				err := <-eventErr
+				// Need to cancel context to close the stream
+				cancel()
+				if err == io.EOF {
```

Comment:
why is EOF okay?

Reply:
According to the Events() specification, once the stream has been completely read an io.EOF error will be sent over the error channel, and it is safe to close the channel by cancelling the context.
 This use case is currently not possible because it relates to a certain set of flags you pass to docker events, --until (stream events until this timestamp) and --since (show all events created since timestamp). Currently we do not pass those flags in to maintain go-dockerclient functionality. In order to use this functionality, we would just need to pass in the requisite flag to the ContainerEvents() function.

Reply:
To keep the same functionality, when you get an EOF error, you can just return instead of reopening, like what fsouza/go-dockerclient does:
https://github.com/fsouza/go-dockerclient/blob/11e1e492616b3d3b7046fceb0f1e7e02f9eef0d8/event.go#L229

Although the EOF error will never happen theoretically.